### PR TITLE
Add set access, multiple harvests and blacklisting to OAIHarvester

### PR DESCRIPTION
**Pull request by no-reply**
- Adds `OAIHarvester#sets` to list sets.
 - Adds ability to specify multiple sets to harvest by passing a list to the `:set` option on `#records` and `#record_ids`.
 - Adds ability to blacklist sets in `#records` and `#record_ids` requests.

### CODE REVIEW

File: lib/krikri/harvesters/oai_harvester.rb [hunk 1]
```diff
@@ -1,6 +1,27 @@
...
+  #          If multiple sets are given, they will be lazily requested from
+  #          `OAI::Client#list_records` in turn and combined into a single
+  #          enumerator.
+  #   - skip_set: A string or array of strings specifying the sets to skip.
+  #               If both `set` and `skip_set` are given, sets specified as
+  #               skip are excluded from the harvest. Otherwise, all sets
+  #               returned by `#set` except those skipped will be harvested.
+  #   - metadata_prefix: A string specifying the metadata prefix. e.g. 'oai_dc'.
+  #   - from: The begin date for the harvest.
+  #   - until: The end date for the harvest.
```
Comment (anarchivist) [inline lib/krikri/harvesters/oai_harvester.rb]: Should there be a `@see` reference to the OAI gem docs?

File: lib/krikri/harvesters/oai_harvester.rb [hunk 2]
```diff
@@ -51,38 +76,107 @@ def count
...
-                            record_xml(rec))
-
+      opts = @opts.merge(opts)
+      request_with_sets(opts) do |set_opts|
+        client.list_records(set_opts).full.lazy.flat_map do |rec|
+          @record_class.build(mint_id(rec.header.identifier),
+                              record_xml(rec))
+        end
       end
+      # sets = Array(opts.delete(:set) { [nil] })
```
Comment (anarchivist) [inline lib/krikri/harvesters/oai_harvester.rb]: Do you still want to retain this commented-out code for some reason?

File: lib/krikri/harvesters/oai_harvester.rb [hunk 3]
```diff
@@ -51,38 +76,107 @@ def count
...
+          until: {type: :string, required: false}
         }
       }
     end
 
+    ##
+    # Concatinates two enumerators
+    # TODO: find a better home for this. Reopen Enumerable?
+    def concat_enum(*enums)
+      Enumerator.new do |yielder|
```
Comment (anarchivist) [inline lib/krikri/harvesters/oai_harvester.rb]: Would it be possible to do something like `enums.flat_map { |e| e }`, or am I missing something about the seeming simplicity of that?

File: spec/lib/krikri/harvesters/oai_harvester_spec.rb [hunk 4]
```diff
@@ -243,6 +378,20 @@
       end
     end
 
+    describe 'concat_enum' do
+
+      it 'concatenaces enums' do
```
Comment (anarchivist) [inline spec/lib/krikri/harvesters/oai_harvester_spec.rb]: typo

**Comment by anarchivist** [comment]
I think this looks good. My most substantive feedback is about `#concat_enums`...

File: lib/krikri/harvesters/oai_harvester.rb (hunk 3 above)
Comment (no-reply) [inline lib/krikri/harvesters/oai_harvester.rb]: I think `enums.flat_map { |e| e }` is a no-op.  I thought `enums.flat_map { |e| e.each { |i| i } }` might work, but that seems also to fail nastily.

File: spec/lib/krikri/harvesters/oai_harvester_spec.rb (hunk 4 above)
Comment (no-reply) [inline spec/lib/krikri/harvesters/oai_harvester_spec.rb]: what excellent spelling you have, @no_reply

**Comment by no-reply** [comment]
I believe this is ready. I'm still not singing the praises of `#concat_enums`... or it living in `OAIHarvester` (or really, our code at all... why isn't this in Ruby?) but it works, and I'm inclined to run with it.

I can move it to its own `ConcatenatedEnumerator` class, if that's desired.

File: lib/krikri/harvesters/oai_harvester.rb (hunk 1 above)
Comment (no-reply) [inline lib/krikri/harvesters/oai_harvester.rb]: Done.

**Comment by anarchivist** [comment]
I don't know if we want to rely on an external gem to do this, but there's also `Enumerating`: https://github.com/mdub/enumerating#lazy-combination-of-enumerables

**Comment by anarchivist** [comment]
But otherwise, I'm good on :squash:ing.

**Comment by no-reply** [comment]
Yeah, I was reluctant to introduce a (unmaintained?) dependency for something so small.  I'm going to add a comment with a reference to that, so if we end up doing more of this concatenation, we can reevaluate.

This is squashed to the level I think is appropriate.

**Pull request merged**